### ISSUE TITLE
lxc list seems slow

### ISSUE THREAD

**Issue by tych0** [lxc/lxd#671]
```
ubuntu@criu1:~$ time lxc list
+------+---------+---------------+-------------------------------------+-----------+-----------+
| NAME |  STATE  |     IPV4      |                IPV6                 | EPHEMERAL | SNAPSHOTS |
+------+---------+---------------+-------------------------------------+-----------+-----------+
| mig  | RUNNING | 192.168.1.206 | fd5d:e5bb:c5f9:0:216:3eff:fe5f:f2ec | NO        | 0         |
|  001 | RUNNING | 10.0.3.116    |                                     | NO        | 0         |
|  002 | RUNNING | 10.0.3.162    |                                     | NO        | 0         |
|  003 | RUNNING | 10.0.3.26     |                                     | NO        | 0         |
|  004 | RUNNING | 10.0.3.55     |                                     | NO        | 0         |
|  005 | RUNNING | 10.0.3.114    |                                     | NO        | 0         |
|  006 | RUNNING | 10.0.3.247    |                                     | NO        | 0         |
|  007 | RUNNING | 10.0.3.238    |                                     | NO        | 0         |
|  008 | RUNNING | 10.0.3.235    |                                     | NO        | 0         |
|  009 | RUNNING | 10.0.3.243    |                                     | NO        | 0         |
|  010 | RUNNING | 10.0.3.196    |                                     | NO        | 0         |
|  011 | RUNNING | 10.0.3.84     |                                     | NO        | 0         |
|  012 | RUNNING | 10.0.3.129    |                                     | NO        | 0         |
|  013 | RUNNING | 10.0.3.252    |                                     | NO        | 0         |
|  014 | RUNNING | 10.0.3.48     |                                     | NO        | 0         |
|  015 | RUNNING | 10.0.3.219    |                                     | NO        | 0         |
|  016 | RUNNING | 10.0.3.73     |                                     | NO        | 0         |
|  017 | RUNNING | 10.0.3.49     |                                     | NO        | 0         |
|  018 | RUNNING | 10.0.3.139    |                                     | NO        | 0         |
|  019 | RUNNING | 10.0.3.177    |                                     | NO        | 0         |
|  020 | RUNNING | 10.0.3.66     |                                     | NO        | 0         |
|  021 | RUNNING | 10.0.3.158    |                                     | NO        | 0         |
|  022 | RUNNING | 10.0.3.189    |                                     | NO        | 0         |
|  023 | RUNNING | 10.0.3.56     |                                     | NO        | 0         |
|  024 | RUNNING | 10.0.3.92     |                                     | NO        | 0         |
|  025 | RUNNING | 10.0.3.57     |                                     | NO        | 0         |
|  026 | RUNNING | 10.0.3.90     |                                     | NO        | 0         |
|  027 | RUNNING | 10.0.3.231    |                                     | NO        | 0         |
|  028 | RUNNING | 10.0.3.34     |                                     | NO        | 0         |
|  029 | RUNNING | 10.0.3.180    |                                     | NO        | 0         |
|  030 | RUNNING | 10.0.3.172    |                                     | NO        | 0         |
|  031 | RUNNING | 10.0.3.58     |                                     | NO        | 0         |
|  032 | RUNNING | 10.0.3.81     |                                     | NO        | 0         |
|  033 | RUNNING | 10.0.3.167    |                                     | NO        | 0         |
|  034 | RUNNING | 10.0.3.95     |                                     | NO        | 0         |
|  035 | RUNNING | 10.0.3.193    |                                     | NO        | 0         |
|  036 | RUNNING | 10.0.3.198    |                                     | NO        | 0         |
|  037 | RUNNING | 10.0.3.97     |                                     | NO        | 0         |
|  038 | RUNNING | 10.0.3.118    |                                     | NO        | 0         |
|  039 | RUNNING | 10.0.3.78     |                                     | NO        | 0         |
|  040 | RUNNING | 10.0.3.79     |                                     | NO        | 0         |
|  041 | RUNNING | 10.0.3.117    |                                     | NO        | 0         |
|  042 | RUNNING | 10.0.3.188    |                                     | NO        | 0         |
|  043 | RUNNING | 10.0.3.234    |                                     | NO        | 0         |
|  044 | RUNNING | 10.0.3.27     |                                     | NO        | 0         |
|  045 | RUNNING | 10.0.3.72     |                                     | NO        | 0         |
|  046 | RUNNING | 10.0.3.151    |                                     | NO        | 0         |
|  047 | RUNNING | 10.0.3.171    |                                     | NO        | 0         |
|  048 | RUNNING | 10.0.3.119    |                                     | NO        | 0         |
|  049 | RUNNING | 10.0.3.99     |                                     | NO        | 0         |
|  050 | RUNNING | 10.0.3.93     |                                     | NO        | 0         |
|  051 | RUNNING | 10.0.3.21     |                                     | NO        | 0         |
|  052 | RUNNING | 10.0.3.233    |                                     | NO        | 0         |
|  053 | RUNNING | 10.0.3.164    |                                     | NO        | 0         |
|  054 | RUNNING | 10.0.3.248    |                                     | NO        | 0         |
|  055 | RUNNING | 10.0.3.220    |                                     | NO        | 0         |
|  056 | RUNNING | 10.0.3.59     |                                     | NO        | 0         |
|  057 | RUNNING | 10.0.3.182    |                                     | NO        | 0         |
|  058 | RUNNING | 10.0.3.156    |                                     | NO        | 0         |
|  059 | RUNNING | 10.0.3.157    |                                     | NO        | 0         |
|  060 | RUNNING | 10.0.3.126    |                                     | NO        | 0         |
|  061 | RUNNING | 10.0.3.138    |                                     | NO        | 0         |
|  062 | RUNNING | 10.0.3.222    |                                     | NO        | 0         |
|  063 | RUNNING | 10.0.3.213    |                                     | NO        | 0         |
|  064 | RUNNING | 10.0.3.228    |                                     | NO        | 0         |
|  065 | RUNNING | 10.0.3.74     |                                     | NO        | 0         |
|  066 | RUNNING | 10.0.3.68     |                                     | NO        | 0         |
|  067 | RUNNING | 10.0.3.148    |                                     | NO        | 0         |
|  068 | RUNNING | 10.0.3.107    |                                     | NO        | 0         |
|  069 | RUNNING | 10.0.3.254    |                                     | NO        | 0         |
|  070 | RUNNING | 10.0.3.75     |                                     | NO        | 0         |
|  071 | RUNNING | 10.0.3.140    |                                     | NO        | 0         |
|  072 | RUNNING | 10.0.3.215    |                                     | NO        | 0         |
|  073 | RUNNING | 10.0.3.168    |                                     | NO        | 0         |
|  074 | RUNNING | 10.0.3.175    |                                     | NO        | 0         |
|  075 | RUNNING | 10.0.3.88     |                                     | NO        | 0         |
|  076 | RUNNING | 10.0.3.91     |                                     | NO        | 0         |
|  077 | RUNNING | 10.0.3.28     |                                     | NO        | 0         |
|  078 | RUNNING | 10.0.3.236    |                                     | NO        | 0         |
|  079 | RUNNING | 10.0.3.121    |                                     | NO        | 0         |
|  080 | RUNNING | 10.0.3.149    |                                     | NO        | 0         |
|  081 | RUNNING | 10.0.3.214    |                                     | NO        | 0         |
|  082 | RUNNING | 10.0.3.98     |                                     | NO        | 0         |
|  083 | RUNNING | 10.0.3.120    |                                     | NO        | 0         |
|  084 | RUNNING | 10.0.3.122    |                                     | NO        | 0         |
|  085 | RUNNING | 10.0.3.237    |                                     | NO        | 0         |
|  086 | RUNNING | 10.0.3.223    |                                     | NO        | 0         |
|  087 | RUNNING | 10.0.3.12     |                                     | NO        | 0         |
|  088 | RUNNING | 10.0.3.105    |                                     | NO        | 0         |
|  089 | RUNNING | 10.0.3.186    |                                     | NO        | 0         |
|  090 | RUNNING | 10.0.3.204    |                                     | NO        | 0         |
|  091 | RUNNING | 10.0.3.100    |                                     | NO        | 0         |
|  092 | RUNNING | 10.0.3.109    |                                     | NO        | 0         |
|  093 | RUNNING | 10.0.3.192    |                                     | NO        | 0         |
|  094 | RUNNING | 10.0.3.70     |                                     | NO        | 0         |
|  095 | RUNNING | 10.0.3.76     |                                     | NO        | 0         |
|  096 | RUNNING | 10.0.3.239    |                                     | NO        | 0         |
|  097 | RUNNING | 10.0.3.77     |                                     | NO        | 0         |
|  098 | RUNNING | 10.0.3.108    |                                     | NO        | 0         |
|  099 | RUNNING | 10.0.3.176    |                                     | NO        | 0         |
|  100 | RUNNING | 10.0.3.123    |                                     | NO        | 0         |
+------+---------+---------------+-------------------------------------+-----------+-----------+

real	0m1.697s
user	0m0.012s
sys	0m0.008s
```

Should it really take 1.7 seconds to do a list on a reasonably powerful (i7) system with only 100 containers?

This is exacerbated on arm systems which don't have great core clockspeeds to begin with.

**Comment by dlezcano** [comment]
On 05/18/2015 01:03 AM, Tycho Andersen wrote:
> |ubuntu@criu1:~$ time lxc list
> +------+---------+---------------+-------------------------------------+-----------+-----------+
> | NAME |  STATE  |     IPV4      |                IPV6                 | EPHEMERAL | SNAPSHOTS |
> +------+---------+---------------+-------------------------------------+-----------+-----------+
> | mig  | RUNNING | 192.168.1.206 | fd5d:e5bb:c5f9:0:216:3eff:fe5f:f2ec | NO        | 0         |
> |  001 | RUNNING | 10.0.3.116    |                                     | NO        | 0         |
> |  002 | RUNNING | 10.0.3.162    |                                     | NO        | 0         |
> |  003 | RUNNING | 10.0.3.26     |                                     | NO        | 0         |
> |  004 | RUNNING | 10.0.3.55     |                                     | NO        | 0         |
> |  005 | RUNNING | 10.0.3.114    |                                     | NO        | 0         |
> |  006 | RUNNING | 10.0.3.247    |                                     | NO        | 0         |
> |  007 | RUNNING | 10.0.3.238    |                                     | NO        | 0         |
> |  008 | RUNNING | 10.0.3.235    |                                     | NO        | 0         |
> |  009 | RUNNING | 10.0.3.243    |                                     | NO        | 0         |
> |  010 | RUNNING | 10.0.3.196    |                                     | NO        | 0         |
> |  011 | RUNNING | 10.0.3.84     |                                     | NO        | 0         |
> |  012 | RUNNING | 10.0.3.129    |                                     | NO        | 0         |
> |  013 | RUNNING | 10.0.3.252    |                                     | NO        | 0         |
> |  014 | RUNNING | 10.0.3.48     |                                     | NO        | 0         |
> |  015 | RUNNING | 10.0.3.219    |                                     | NO        | 0         |
> |  016 | RUNNING | 10.0.3.73     |                                     | NO        | 0         |
> |  017 | RUNNING | 10.0.3.49     |                                     | NO        | 0         |
> |  018 | RUNNING | 10.0.3.139    |                                     | NO        | 0         |
> |  019 | RUNNING | 10.0.3.177    |                                     | NO        | 0         |
> |  020 | RUNNING | 10.0.3.66     |                                     | NO        | 0         |
> |  021 | RUNNING | 10.0.3.158    |                                     | NO        | 0         |
> |  022 | RUNNING | 10.0.3.189    |                                     | NO        | 0         |
> |  023 | RUNNING | 10.0.3.56     |                                     | NO        | 0         |
> |  024 | RUNNING | 10.0.3.92     |                                     | NO        | 0         |
> |  025 | RUNNING | 10.0.3.57     |                                     | NO        | 0         |
> |  026 | RUNNING | 10.0.3.90     |                                     | NO        | 0         |
> |  027 | RUNNING | 10.0.3.231    |                                     | NO        | 0         |
> |  028 | RUNNING | 10.0.3.34     |                                     | NO        | 0         |
> |  029 | RUNNING | 10.0.3.180    |                                     | NO        | 0         |
> |  030 | RUNNING | 10.0.3.172    |                                     | NO        | 0         |
> |  031 | RUNNING | 10.0.3.58     |                                     | NO        | 0         |
> |  032 | RUNNING | 10.0.3.81     |                                     | NO        | 0         |
> |  033 | RUNNING | 10.0.3.167    |                                     | NO        | 0         |
> |  034 | RUNNING | 10.0.3.95     |                                     | NO        | 0         |
> |  035 | RUNNING | 10.0.3.193    |                                     | NO        | 0         |
> |  036 | RUNNING | 10.0.3.198    |                                     | NO        | 0         |
> |  037 | RUNNING | 10.0.3.97     |                                     | NO        | 0         |
> |  038 | RUNNING | 10.0.3.118    |                                     | NO        | 0         |
> |  039 | RUNNING | 10.0.3.78     |                                     | NO        | 0         |
> |  040 | RUNNING | 10.0.3.79     |                                     | NO        | 0         |
> |  041 | RUNNING | 10.0.3.117    |                                     | NO        | 0         |
> |  042 | RUNNING | 10.0.3.188    |                                     | NO        | 0         |
> |  043 | RUNNING | 10.0.3.234    |                                     | NO        | 0         |
> |  044 | RUNNING | 10.0.3.27     |                                     | NO        | 0         |
> |  045 | RUNNING | 10.0.3.72     |                                     | NO        | 0         |
> |  046 | RUNNING | 10.0.3.151    |                                     | NO        | 0         |
> |  047 | RUNNING | 10.0.3.171    |                                     | NO        | 0         |
> |  048 | RUNNING | 10.0.3.119    |                                     | NO        | 0         |
> |  049 | RUNNING | 10.0.3.99     |                                     | NO        | 0         |
> |  050 | RUNNING | 10.0.3.93     |                                     | NO        | 0         |
> |  051 | RUNNING | 10.0.3.21     |                                     | NO        | 0         |
> |  052 | RUNNING | 10.0.3.233    |                                     | NO        | 0         |
> |  053 | RUNNING | 10.0.3.164    |                                     | NO        | 0         |
> |  054 | RUNNING | 10.0.3.248    |                                     | NO        | 0         |
> |  055 | RUNNING | 10.0.3.220    |                                     | NO        | 0         |
> |  056 | RUNNING | 10.0.3.59     |                                     | NO        | 0         |
> |  057 | RUNNING | 10.0.3.182    |                                     | NO        | 0         |
> |  058 | RUNNING | 10.0.3.156    |                                     | NO        | 0         |
> |  059 | RUNNING | 10.0.3.157    |                                     | NO        | 0         |
> |  060 | RUNNING | 10.0.3.126    |                                     | NO        | 0         |
> |  061 | RUNNING | 10.0.3.138    |                                     | NO        | 0         |
> |  062 | RUNNING | 10.0.3.222    |                                     | NO        | 0         |
> |  063 | RUNNING | 10.0.3.213    |                                     | NO        | 0         |
> |  064 | RUNNING | 10.0.3.228    |                                     | NO        | 0         |
> |  065 | RUNNING | 10.0.3.74     |                                     | NO        | 0         |
> |  066 | RUNNING | 10.0.3.68     |                                     | NO        | 0         |
> |  067 | RUNNING | 10.0.3.148    |                                     | NO        | 0         |
> |  068 | RUNNING | 10.0.3.107    |                                     | NO        | 0         |
> |  069 | RUNNING | 10.0.3.254    |                                     | NO        | 0         |
> |  070 | RUNNING | 10.0.3.75     |                                     | NO        | 0         |
> |  071 | RUNNING | 10.0.3.140    |                                     | NO        | 0         |
> |  072 | RUNNING | 10.0.3.215    |                                     | NO        | 0         |
> |  073 | RUNNING | 10.0.3.168    |                                     | NO        | 0         |
> |  074 | RUNNING | 10.0.3.175    |                                     | NO        | 0         |
> |  075 | RUNNING | 10.0.3.88     |                                     | NO        | 0         |
> |  076 | RUNNING | 10.0.3.91     |                                     | NO        | 0         |
> |  077 | RUNNING | 10.0.3.28     |                                     | NO        | 0         |
> |  078 | RUNNING | 10.0.3.236    |                                     | NO        | 0         |
> |  079 | RUNNING | 10.0.3.121    |                                     | NO        | 0         |
> |  080 | RUNNING | 10.0.3.149    |                                     | NO        | 0         |
> |  081 | RUNNING | 10.0.3.214    |                                     | NO        | 0         |
> |  082 | RUNNING | 10.0.3.98     |                                     | NO        | 0         |
> |  083 | RUNNING | 10.0.3.120    |                                     | NO        | 0         |
> |  084 | RUNNING | 10.0.3.122    |                                     | NO        | 0         |
> |  085 | RUNNING | 10.0.3.237    |                                     | NO        | 0         |
> |  086 | RUNNING | 10.0.3.223    |                                     | NO        | 0         |
> |  087 | RUNNING | 10.0.3.12     |                                     | NO        | 0         |
> |  088 | RUNNING | 10.0.3.105    |                                     | NO        | 0         |
> |  089 | RUNNING | 10.0.3.186    |                                     | NO        | 0         |
> |  090 | RUNNING | 10.0.3.204    |                                     | NO        | 0         |
> |  091 | RUNNING | 10.0.3.100    |                                     | NO        | 0         |
> |  092 | RUNNING | 10.0.3.109    |                                     | NO        | 0         |
> |  093 | RUNNING | 10.0.3.192    |                                     | NO        | 0         |
> |  094 | RUNNING | 10.0.3.70     |                                     | NO        | 0         |
> |  095 | RUNNING | 10.0.3.76     |                                     | NO        | 0         |
> |  096 | RUNNING | 10.0.3.239    |                                     | NO        | 0         |
> |  097 | RUNNING | 10.0.3.77     |                                     | NO        | 0         |
> |  098 | RUNNING | 10.0.3.108    |                                     | NO        | 0         |
> |  099 | RUNNING | 10.0.3.176    |                                     | NO        | 0         |
> |  100 | RUNNING | 10.0.3.123    |                                     | NO        | 0         |
> +------+---------+---------------+-------------------------------------+-----------+-----------+
>
> real    0m1.697s
> user    0m0.012s
> sys 0m0.008s
> |
>
> Should it really take 1.7 seconds to do a list on a reasonably powerful
> (i7) system with only 100 containers?
>
> This is exacerbated on arm systems which don't have great core
> clockspeeds to begin with.

It would be interesting to show the current cpu usage.

As far as I can see, the user and sys time are very short but the clock 
wall is longer. So very likely you a lot of processes competing for cpu 
time, hence smaller time slices to each processes and longer time to 
process a command.

**Comment by stgraber** [comment]
Was that using current git master?

We landed support for ?recursion=1 very recently for /1.0/containers, so instead of having to do one query per container, a single GET can now return everything in one shot.

If current git master is still slow, then it's likely the all the attach operations that are taking a while, though in my experience using straight LXC on pretty powerful systems, I can list 200-300 containers in around 30s (would have to re-check on recent LXC though).

**Comment by tych0** [comment]
Yes, this was with the recursion patch.

**Comment by tych0** [comment]
On Mon, May 18, 2015 at 01:46:23AM -0700, Daniel Lezcano wrote:
> 
> It would be interesting to show the current cpu usage.
> 
> As far as I can see, the user and sys time are very short but the clock 
> wall is longer. So very likely you a lot of processes competing for cpu 
> time, hence smaller time slices to each processes and longer time to 
> process a command.

The time numbers there are just for the client, which is just waiting
for a http response for the server. I straced the server just to see
if it was blocking anywhere and didn't, so I suspect it might be a
worst case of "general slowness because it's doing a lot of stuff" :(

**Comment by stgraber** [comment]
Might be worth comparing the speed of lxc list vs lxc-ls -f, if the latter is somehow faster, then we should go look into go-lxc's implementation of get_ips at least.

**Comment by stgraber** [comment]
When serving a recursive query, we may also be able to do the get_ips and other data gathering in parallel inside go-routines, maybe something to consider.

**Comment by mikemccracken** [comment]
as an added data point, with 499 containers on my i5 NUC, 'lxc list' took 9.6 minutes…

**Comment by pcdummy** [comment]
I tried to implement #875 - goroutines for that, sometimes its slower sometimes as fast as serial data gathering. I don't know enough about go-lxc to improve that any further.

**Comment by jensg-st** [comment]
I reckon you have figured out, what the issue is by now, but I had the same problem, so I had a look. I used 83 containers. A list takes me between 800ms - 1.2s.
First I merged the SQL statement, so I'm just using one instead of one per container.  That saved 50 - 100ms, so not a big deal. Looking further I figured out, that getting the IPs for each container is the part where the time gets lost. After removing the part:

```
	pid, _ := c.InitPidGet()
	status.Init = pid
	status.Ips = c.iPsGet()
```

from RenderState() from the containers, it drops down to 200ms including the DB query. I still think that the db query should be merged, but that is not a big win although it would scale better.

There are is a simple solution: Removing IPs from the container list API call or add a flag, where you can specify to get the IPs as well. And on the client, by default, there are no IPs in the table. Just with --with-ips or something.

Another solution would be to replace the full IP information with just strings of IPs in lists. The call 'IPAddresses()' takes only ~30% of the time on my machine compared to the 'iPsGet' implementation.

But again, maybe you have found a solution already.

**Comment by srkunze** [comment]
Confirmed it for lxd-0.23 and 100 containers: roughly 3.3 seconds. I would expect not even a millisecond.

**Comment by srkunze** [comment]
Those 100 containers not running: 0.6 seconds. Seems still too much to me.

**Comment by tych0** [comment]
Ok, I'm going to take a hard look at this again a bit before the holiday break and see if I can figure out wtf is going on.

**Comment by stgraber** [comment]
Right so I made two changes to improve the situation here, not that the user visible end result is unfortunately similar though.

I rewrote the network handling code to be in a separate process, meaning that parallel queries now improve performance rather than just clog the server.

We've moved the container status information under a separate API call, the client then batches those calls by batch of 10 in parallel to go through them faster.

The problem though is that we've also now added cgroup and storage information to the mix which itself takes quite a while for the server to retrieve, so the end result is slightly less than 100ms per container, which isn't exactly impressive.

However if you don't care about that data (network info and snapshots), then using "lxc list -c nse" is pretty fast at about 8ms per container (when dealing with a large set).
